sa/database.go

```diff
@@ -156,12 +160,173 @@ func NewDbMapFromConfig(config *mysql.Config, settings DbSettings) (*boulderDB.W
+		"session_track_system_variables": {},
+		"shared_memory_base_name":        {},
+		"sql_mode":                       {},
+		"system_time_zone":               {},
+		"time_zone":                      {},
+		"version":                        {},
+		"version_comment":                {},
+		"version_compile_machine":        {},
+		"version_compile_os":             {},
+		"version_malloc_library":         {},
+		"version_source_revision":        {},
+	}
+
+	// Keep in mind that certain MariaDB enums can have both string and integer
+	// values.
+	mariaDBStringEnumTypes := map[string]struct{}{
+		"completion_type":                {},
+		"concurrent_insert":              {},
+		"default_storage_engine":         {},
+		"default_tmp_storage_engine":     {},
+		"delay_key_write":                {},
+		"event_scheduler":                {},
+		"optimizer_trace":                {},
+		"plugin_maturity":                {},
+		"session_track_transaction_info": {},
+		"tx_isolation":                   {},
+	}
+
+	// Keep in mind that certain MariaDB enums can have both string and integer
+	// values.
+	mariaDBIntEnumTypes := map[string]struct{}{
+		"completion_type":   {},
+		"concurrent_insert": {},
+		"event_scheduler":   {},
+	}
+
+	// Check the list of currently known MariaDB string type system variables
+	// and determine if the value is a properly formatted string e.g.
+	// sql_mode='STRICT_TABLES'
+	if _, found := mariaDBSessionScope[name]; !found {
+		return fmt.Errorf("%v is unable to be used in the session scope", name)
+	}
+
+	// Check the list of currently known MariaDB string type system variables
+	// and determine if the value is a properly formatted string e.g.
+	// sql_mode='STRICT_TABLES'
+	if _, found := mariaDBStringTypes[name]; found {
+		if checkStringQuoteRE.FindString(value) == value {
+			return nil
+		}
+		return fmt.Errorf("%v=%v string is not properly quoted", name, value)
+	}
+
+	// Check the list of currently known MariaDB enumeration type system
+	// variables and determine if the value is either:
+	// 1) A properly formatted integer e.g. completion_type=1
+	if _, found := mariaDBIntEnumTypes[name]; found {
+		if checkIntRE.FindString(value) == value {
+			return nil
+		} else if checkImproperIntRE.FindString(value) == value {
+			return fmt.Errorf("%v=%v enumeration as integer is quoted, but should not be", name, value)
+		}
+	}
+
+	// 2) A properly formatted string e.g. completion_type='CHAIN'
+	if _, found := mariaDBStringEnumTypes[name]; found {
+		if checkStringQuoteRE.FindString(value) == value {
+			return nil
+		}
+		return fmt.Errorf("%v=%v enumeration as string is not properly quoted", name, value)
+	}
+
+	// Finally, we don't perform any server variable validation checks against MariaDB
+	// booleans or numerics (which can be either integers or floats)
+	// each of which are represented as a Golang string type.
+	if checkFloatOrIntRE.FindString(value) == value || value == "true" || value == "false" {
+		return nil
+	}
```

Comment:
I don't think this clause is necessary. We only want to check if we're quoting string values. If it's not a string or a string or integer valued enumeration then we can simply exit (nil) early.

Reply:
FWIW (as of MariaDB 5.6.2) system variable boolean values are allowed to be 0, 1, OFF, ON, FALSE, TRUE.

Reply:
Good catch on the OFF, ON. I've since removed that check on line 321 and rewrote nearly this entire function.